internal/shell/shell.go

```diff
@@ -256,6 +264,34 @@ func (s *Shell) exec(ctx context.Context, command string) (string, string, error
 	return stdout.String(), stderr.String(), err
 }
 
+// execPOSIXStream executes commands using POSIX shell emulation with streaming output
+func (s *Shell) execPOSIXStream(ctx context.Context, command string, stdout, stderr io.Writer) error {
+	line, err := syntax.NewParser().Parse(strings.NewReader(command), "")
+	if err != nil {
+		return fmt.Errorf("could not parse command: %w", err)
+	}
+
+	runner, err := interp.New(
+		interp.StdIO(nil, stdout, stderr),
+		interp.Interactive(false),
+		interp.Env(expand.ListEnviron(s.env...)),
+		interp.Dir(s.cwd),
+		interp.ExecHandlers(s.execHandlers()...),
+	)
+	if err != nil {
+		return fmt.Errorf("could not run command: %w", err)
+	}
+
+	err = runner.Run(ctx, line)
+	s.cwd = runner.Dir
+	s.env = []string{}
+	for name, vr := range runner.Vars {
+		s.env = append(s.env, fmt.Sprintf("%s=%s", name, vr.Str))
+	}
+	s.logger.InfoPersist("POSIX command finished", "command", command, "err", err)
+	return err
```

Comment:
Looks like 90%+ of this code is duplicated between `exec` and `execStream`.

Create a separate `newInterp` function to de-duplicate this.